File: datastructure/BinaryTreeNode.go
package datastructure

import (
	"strconv"
)

// BinaryTreeNode literally indicates a node of a binary tree.
type BinaryTreeNode struct {
	Val   int
	Left  *BinaryTreeNode
	Right *BinaryTreeNode
}
type queueData struct {
	node   *BinaryTreeNode
	isLeft bool
}

// CreateBinaryTreeNodeFromString creates a tructure of BinaryTreeNode
// from an array of string(or "null") in LeetCode expression
// ex) ["1", "2", "null", "3"]
func CreateBinaryTreeNodeFromString(numStrings []string) *BinaryTreeNode {
	if len(numStrings) == 0 {
		return nil
	}

	numString := numStrings[0]
	numStrings = numStrings[1:]

	num, err := strconv.Atoi(numString)

	if err != nil {
		return nil
	}

	root := BinaryTreeNode{num, nil, nil}

	if len(numStrings) == 0 {
		return &root
	}

	queue := []queueData{
		queueData{&root, true},
		queueData{&root, false},
	}

	for len(numStrings) > 0 {
		q := queue[0]
		queue = queue[1:]
		numString := numStrings[0]
		numStrings = numStrings[1:]

		num, err := strconv.Atoi(numString)

		if err != nil {
			continue
		}

		if q.isLeft {
			q.node.Left = &BinaryTreeNode{num, nil, nil}
			queue = append(
				queue,
				queueData{q.node.Left, true},
				queueData{q.node.Left, false},
			)
		} else {
			q.node.Right = &BinaryTreeNode{num, nil, nil}
			queue = append(
				queue,
				queueData{q.node.Right, true},
				queueData{q.node.Right, false},
			)
		}
	}

	return &root
}

File: datastructure/BinaryTreeNode.ts
export interface BinaryTreeNode<T> {
  val: T;
  left: BinaryTreeNode<T> | null;
  right: BinaryTreeNode<T> | null;
}

type isLeft = boolean;

export function createBinaryTreeNode<T>(
  values: (T | null)[]
): BinaryTreeNode<T> | null {
  if (values.length === 0) return null;

  const root = createSingleNode(values[0]);
  let queue: [BinaryTreeNode<T>, isLeft][] = [
    [root!, true],
    [root!, false],
  ];

  for (let i = 1; i < values.length; i++) {
    const [node, isLeft] = queue.shift()!;
    const childNode = createSingleNode(values[i]);
    if (isLeft) {
      node.left = childNode;
    } else {
      node.right = childNode;
    }
    if (childNode) {
      queue.push([childNode, true], [childNode, false]);
    }
  }

  return root;
}

function createSingleNode<T>(value: T | null): BinaryTreeNode<T> | null {
  return value === null ? null : { val: value, left: null, right: null };
}
